jobs/fix_crt_channel_unreads/worker.go

```diff
@@ -0,0 +1,260 @@
+		close(w.stoppedChan)
+	}()
+
+	for {
+		select {
+		case <-w.stopChan:
+			mlog.Debug("Worker received stop signal", mlog.String("worker", w.name))
+			return
+		case job := <-w.jobsChan:
+			mlog.Debug("Worker received a new candidate job.", mlog.String("worker", w.name))
+			w.doJob(&job)
+		}
+	}
+}
+
+func (w *FixCRTChannelUnreadsWorker) Stop() {
+	mlog.Debug("Worker stopping", mlog.String("worker", w.name))
+	close(w.stopChan)
+	<-w.stoppedChan
+}
+
+func (w *FixCRTChannelUnreadsWorker) doJob(job *model.Job) {
+	if claimed, err := w.jobServer.ClaimJob(job); err != nil {
+		mlog.Warn("Worker experienced an error while trying to claim job",
+			mlog.String("worker", w.name),
+			mlog.String("job_id", job.Id),
+			mlog.String("error", err.Error()))
+		return
+	} else if !claimed {
+		return
+	}
+	if _, err := w.app.Srv().Store.System().GetByName(model.MigrationKeyFixCRTChannelUnreads); err == nil {
+		mlog.Info("Worker: migration already done", mlog.String("worker", w.name), mlog.String("job_id", job.Id))
+		w.setJobSuccess(job)
+		return
+	}
+
+	fixedBadCM := 0
+	migrationDone := false
+	prevErr := false
+	userID, channelID := w.getProgressFromPreviousJobs(job)
+	if userID != "" && channelID != "" {
+		mlog.Info("Restarting from previous job run", mlog.String("UserID", userID), mlog.String("ChannelID", channelID))
+	}
+	shouldStop := false
+	for {
+		select {
+		case <-w.stopChan:
+			shouldStop = true
+		default:
+			shouldStop = false
+		}
+		if shouldStop {
+			break
+		}
+		cms, sErr := w.app.Srv().Store.Channel().GetCRTUnfixedChannelMembershipsAfter(userID, channelID, 100)
+		if sErr != nil {
+			if sErr == sql.ErrNoRows {
+				migrationDone = true
+				break
+			}
+			mlog.Warn("Failed to get bad channel memberships",
+				mlog.String("worker", w.name),
+				mlog.String("job_id", job.Id),
+				mlog.String("error", sErr.Error()))
+			if prevErr {
+				w.updateProgress(job, userID, channelID)
+				w.setJobError(job, model.NewAppError(w.name, "fix_crt_channel_unreads.worker.do_job.get_bad_channel_memberships", nil, sErr.Error(), http.StatusInternalServerError))
+				return
+			}
+			prevErr = true
+			continue
+		}
+		if len(cms) == 0 {
+			migrationDone = true
+			break
+		}
+		lastCM := cms[len(cms)-1]
+		userID = lastCM.UserId
+		channelID = lastCM.ChannelId
+
+		prevErr = false
+		for _, cm := range cms {
+			postTypes, err := w.app.Srv().Store.Post().GetUniquePostTypesSince(cm.ChannelId, cm.LastViewedAt)
+			if err != nil {
+				mlog.Warn("Failed to get unique unread posts",
+					mlog.String("worker", w.name),
+					mlog.String("job_id", job.Id),
+					mlog.String("error", err.Error()))
+				if prevErr {
+					w.updateProgress(job, cm.UserId, cm.ChannelId)
+					w.setJobError(job, model.NewAppError(w.name, "fix_crt_channel_unreads.worker.do_job.get_post_types", nil, err.Error(), http.StatusInternalServerError))
+					return
+				}
+				prevErr = true
+				continue
+			}
+			if !containsNormalPost(postTypes) {
```

Comment:
I am clearly missing some context. Why do we need to act here if no "normal" posts is found?

Reply:
When we added the `TotalMsgCountRoot` (to Channels) and `MsgCountRoot` (to channelMembers) columns, system messages were included in the count (they normally aren't). This lead to many channels being marked as unread for users. These channels only had join/leave messages, and no user created posts. 

This migration fixes that issue by marking a channel as read if there are no user posted messages. At the moment this is a bit heavy handed - it ignores all system messages, not just join/leave messages (i.e. containsNormalPost is `true` only if `post.Type == ""` ). I'm looking into refining this further, but the check will be similar.